include/rmm/mr/device/limiting_resource_adaptor.hpp

```diff
@@ -0,0 +1,221 @@
+   * the underlying device. The device may not be able to support this limit.
+   *
+   * @return std::size_t max number of bytes allowed for this allocator
+   */
+  std::size_t get_allocation_limit() const { return allocation_limit_; }
+
+ private:
+  /**
+   * @brief Allocates memory of size at least `bytes` using the upstream
+   * resource as long as it fits inside the allocation limit.
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   * by the upstream resource.
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param stream Stream on which to perform the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override
+  {
+    void* p = nullptr;
+
+    std::size_t proposed_size = rmm::detail::align_up(bytes, allocation_alignment_);
+    if (proposed_size + allocated_bytes_ <= allocation_limit_) {
+      p = upstream_->allocate(bytes, stream);
+      allocated_bytes_ += proposed_size;
+    } else {
+      throw rmm::bad_alloc{std::string{"CUDA error: Unable to allocate memory due to limit"}};
+    }
+
+    return p;
+  }
+
+  /**
+   * @brief Free allocation of size `bytes` pointed to by `p`
+   *
+   * @throws Nothing.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes Size of the allocation
+   * @param stream Stream on which to perform the deallocation
+   */
+  void do_deallocate(void* p, std::size_t bytes, cudaStream_t stream) override
+  {
+    std::size_t allocated_size = rmm::detail::align_up(bytes, allocation_alignment_);
+    upstream_->deallocate(p, bytes, stream);
+    allocated_bytes_ -= allocated_size;
+  }
+
+  /**
+   * @brief Compare the upstream resource to another.
+   *
+   * @throws Nothing.
+   *
+   * @param other The other resource to compare to
+   * @return true If the two resources are equivalent
+   * @return false If the two resources are not equal
+   */
+  bool do_is_equal(device_memory_resource const& other) const noexcept override
+  {
+    if (this == &other)
+      return true;
+    else {
+      limiting_resource_adaptor<Upstream> const* cast =
+        dynamic_cast<limiting_resource_adaptor<Upstream> const*>(&other);
+      if (cast != nullptr)
+        return upstream_->is_equal(*cast->get_upstream());
+      else
+        return upstream_->is_equal(other);
+    }
+  }
+
+  /**
+   * @brief Get free and available memory from upstream resource.
+   *
+   * @throws `rmm::cuda_error` if unable to retrieve memory info.
+   *
+   * @param stream Stream on which to get the mem info.
+   * @return std::pair contaiing free_size and total_size of memory
+   */
+  std::pair<size_t, size_t> do_get_mem_info(cudaStream_t stream) const override
+  {
+    auto ret = upstream_->get_mem_info(stream);
+    return {std::min(ret.first, allocation_limit_ - allocated_bytes_),
```

Comment:
This logic doesn't look right. Either it should return what the upstream returns, or only return `allocation_limit - allocated_bytes` and `allocation_limit`. Probably the latter.

Reply:
The thought was if you are limited to 8 gigs, but only have a card with 4 gigs that returning 8 gigs was not correct. I'm happy either way though.

Reply:
Good point. But what if `upstream` does not support `get_mem_info`?